File: app/src/main/java/code/name/monkey/retromusic/loaders/ArtistLoader.java
package code.name.monkey.retromusic.loaders;

import android.content.Context;
import android.provider.MediaStore.Audio.AudioColumns;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;

import code.name.monkey.retromusic.model.Album;
import code.name.monkey.retromusic.model.Artist;
import code.name.monkey.retromusic.util.PreferenceUtil;
import io.reactivex.Observable;

public class ArtistLoader {
    private static String getSongLoaderSortOrder(Context context) {
        return PreferenceUtil.getInstance().getArtistSortOrder() + ", " +
                PreferenceUtil.getInstance().getArtistAlbumSortOrder() + ", " +
                PreferenceUtil.getInstance().getAlbumDetailSongSortOrder() + ", " +
                PreferenceUtil.getInstance().getArtistDetailSongSortOrder();
    }

    @NonNull
    public static Observable<Artist> getArtist(@NonNull final Context context, int artistId) {
        return Observable.create(e -> SongLoader.getSongs(SongLoader.makeSongCursor(
                context,
                AudioColumns.ARTIST_ID + "=?",
                new String[]{String.valueOf(artistId)},
                getSongLoaderSortOrder(context)))
                .subscribe(songs -> {
                    Artist artist = new Artist(AlbumLoader.splitIntoAlbums(songs));
                    e.onNext(artist);
                    e.onComplete();
                }));
    }

    @NonNull
    public static Observable<ArrayList<Artist>> getAllArtists(@NonNull final Context context) {
        return Observable.create(e -> SongLoader
                .getSongs(SongLoader.makeSongCursor(
                        context,
                        null,
                        null,
                        getSongLoaderSortOrder(context))
                ).subscribe(songs -> {
                    e.onNext(splitIntoArtists(AlbumLoader.splitIntoAlbums(songs)));
                    e.onComplete();
                }));

    }

    @NonNull
    public static Observable<ArrayList<Artist>> getArtists(@NonNull final Context context, String query) {
        return Observable.create(e -> SongLoader.getSongs(SongLoader.makeSongCursor(
                context,
                AudioColumns.ARTIST + " LIKE ?",
                new String[]{"%" + query + "%"},
                getSongLoaderSortOrder(context))
        ).subscribe(songs -> {
            e.onNext(splitIntoArtists(AlbumLoader.splitIntoAlbums(songs)));
            e.onComplete();
        }));
    }

    @NonNull
    public static ArrayList<Artist> splitIntoArtists(@Nullable final ArrayList<Album> albums) {
        ArrayList<Artist> artists = new ArrayList<>();
        if (albums != null) {
            for (Album album : albums) {
                getOrCreateArtist(artists, album.getArtistId()).albums.add(album);
            }
        }
        return artists;
    }

    private static Artist getOrCreateArtist(ArrayList<Artist> artists, int artistId) {
        for (Artist artist : artists) {
            if (!artist.albums.isEmpty() && !artist.albums.get(0).songs.isEmpty() && artist.albums.get(0).songs.get(0).artistId == artistId) {
                return artist;
            }
        }
        Artist album = new Artist();
        artists.add(album);
        return album;
    }

    public static Observable<ArrayList<Artist>> splitIntoArtists(Observable<ArrayList<Album>> albums) {
        return Observable.create(e -> {
            ArrayList<Artist> artists = new ArrayList<>();
            albums.subscribe(localAlbums -> {
                if (localAlbums != null) {
                    for (Album album : localAlbums) {
                        getOrCreateArtist(artists, album.getArtistId()).albums.add(album);
                    }
                }
                e.onNext(artists);
                e.onComplete();
            });
        });
    }
}

File: app/src/main/java/code/name/monkey/retromusic/loaders/ArtistLoader.kt
/*
 * Copyright (c) 2019 Hemanth Savarala.
 *
 * Licensed under the GNU General Public License v3
 *
 * This is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by
 *  the Free Software Foundation either version 3 of the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 */

package code.name.monkey.retromusic.loaders

import android.content.Context
import android.provider.MediaStore.Audio.AudioColumns
import code.name.monkey.retromusic.model.Album
import code.name.monkey.retromusic.model.Artist
import code.name.monkey.retromusic.util.PreferenceUtilKT

object ArtistLoader {
    private fun getSongLoaderSortOrder(): String {
        return PreferenceUtilKT.artistSortOrder + ", " +
                PreferenceUtilKT.artistAlbumSortOrder + ", " +
                PreferenceUtilKT.artistSongSortOrder
    }

    fun getAllArtists(context: Context): ArrayList<Artist> {
        val songs = SongLoader.getSongs(
            SongLoader.makeSongCursor(
                context,
                null, null,
                getSongLoaderSortOrder()
            )
        )
        return splitIntoArtists(AlbumLoader.splitIntoAlbums(songs))
    }

    fun getArtists(context: Context, query: String): ArrayList<Artist> {
        val songs = SongLoader.getSongs(
            SongLoader.makeSongCursor(
                context,
                AudioColumns.ARTIST + " LIKE ?",
                arrayOf("%$query%"),
                getSongLoaderSortOrder()
            )
        )
        return splitIntoArtists(AlbumLoader.splitIntoAlbums(songs))
    }

    fun splitIntoArtists(albums: ArrayList<Album>?): ArrayList<Artist> {
        val artists = ArrayList<Artist>()
        if (albums != null) {
            for (album in albums) {
                getOrCreateArtist(artists, album.artistId).albums!!.add(album)
            }
        }
        return artists
    }

    private fun getOrCreateArtist(artists: ArrayList<Artist>, artistId: Int): Artist {
        for (artist in artists) {
            if (artist.albums!!.isNotEmpty() && artist.albums[0].songs!!.isNotEmpty() && artist.albums[0].songs!![0].artistId == artistId) {
                return artist
            }
        }
        val album = Artist()
        artists.add(album)
        return album
    }

    @JvmStatic
    fun getArtist(context: Context, artistId: Int): Artist {
        val songs = SongLoader.getSongs(
            SongLoader.makeSongCursor(
                context,
                AudioColumns.ARTIST_ID + "=?",
                arrayOf(artistId.toString()),
                getSongLoaderSortOrder()
            )
        )
        return Artist(AlbumLoader.splitIntoAlbums(songs))
    }
}
